go/src/koding/kites/kloud/credential/mongo.go

```diff
@@ -106,3 +108,470 @@ func (db *mongoStore) Put(_ string, creds map[string]interface{}) error {
+		// Fetch credentials that user has access to.
+		userOnly := modelhelper.Selector{
+			"targetName": "JCredential",
+			"sourceId": bson.M{
+				"$in": ids,
+			},
+			"as": "user",
+		}
+
+		if err := db.fetchCreds(f, acc, user, teams, userOnly, &creds); err != nil {
+			return nil, err
+		}
+
+		log.Debug("fetch user shared credentials: %+v", creds)
+	}
+
+	return creds, nil
+}
+
+func (db *mongoDatabase) fetchCreds(f *Filter, acc *models.Account, user *models.User,
+	teams map[bson.ObjectId]*models.Group, belongs modelhelper.Selector, creds *[]*Cred) error {
+
+	db.log().Debug("fetching credentials for %+v", belongs)
+
+	rels, err := modelhelper.GetAllRelationships(belongs)
+	if err != nil && err != mgo.ErrNotFound {
+		return models.ResError(err, "jRelationship")
+	}
+
+	for _, rel := range rels {
+		cred, err := modelhelper.GetCredentialByID(rel.TargetId)
+		if err == mgo.ErrNotFound {
+			// dangling jRelationship, ignore
+			continue
+		}
+		if err != nil {
+			return models.ResError(err, "jCredential")
+		}
+
+		if f.Provider != "" && cred.Provider != f.Provider {
+			// provider does not match, filter out
+			continue
+		}
+
+		c := &Cred{
+			Ident:    cred.Identifier,
+			Provider: cred.Provider,
+			Title:    cred.Title,
+			Perm: &MongoPerm{
+				Acc:   acc,
+				User:  user,
+				Cred:  cred,
+				Roles: UserRole,
+			},
+		}
+
+		db.log().Debug("fetched %+v", c)
+
+		if team, ok := teams[rel.SourceId]; ok {
+			c.Team = team.Slug
+			c.Perm.(*MongoPerm).Team = team
+		}
+
+		*creds = append(*creds, c)
+	}
+
+	return nil
+}
+
+func (db *mongoDatabase) SetCred(c *Cred) error {
+	f, ok := c.Perm.(*Filter)
+	if !ok {
+		return errors.New("invalid credential permission")
+	}
+
+	perm, err := db.Validate(f, c)
+	if err == nil {
+		c.Perm = perm
+		return nil
+	}
+
+	if !models.IsNotFound(err, "jCredential") {
+		return err
+	}
+
+	mPerm, ok := perm.(*MongoPerm)
+	if !ok {
+		return fmt.Errorf("unable to create credential: %s", err)
+	}
+
+	if mPerm.Acc == nil {
+		return fmt.Errorf("unable to create credential: missing %q account", f.User)
+	}
+
+	now := time.Now().UTC()
+
+	if c.Title == "" {
+		c.Title = mPerm.User.Name + " " + now.String()
+	}
+
+	mPerm.Cred = &models.Credential{
+		Id:          bson.NewObjectId(),
+		Provider:    c.Provider,
+		Identifier:  c.Ident,
+		Title:       c.Title,
+		OriginId:    mPerm.Acc.Id,
+		Verified:    false,
+		AccessLevel: "private",
+		Meta: &models.CredentialMeta{
+			CreatedAt:  now,
+			ModifiedAt: now,
+		},
+	}
+
+	if err := modelhelper.CreateCredential(mPerm.Cred); err != nil {
+		return err
+	}
+
+	rel := &models.Relationship{
+		Id:         bson.NewObjectId(),
+		TargetId:   mPerm.Cred.Id,
+		TargetName: "JCredential",
+		SourceId:   mPerm.Acc.Id,
+		SourceName: "JAccount",
+		As:         "owner",
+		TimeStamp:  now,
+	}
+
+	if err := modelhelper.AddRelationship(rel); err != nil {
+		return err
+	}
+
+	c.Perm = mPerm
+
+	return nil
+}
+
+func (db *mongoDatabase) Lock(cred *Cred) error {
```

Comment:
`create`, `edit` and `remove` operations on a single credential are exclusive ones - while one is in progress, any other should be rejected.

I coded for future, since right now I need only create method for `kd stack create`. I've removed the locking and will re-introduce once other methods are needed.